File: calc/calc_power.py
import sys

from parse.parse_fitfile import parse_fitfile
from pylab import *

def calc_direction(Alat, Along, Blat, Blong):
    Alat_rad = Alat/180.*pi
    Along_rad = Along/180.*pi
    Blat_rad = Blat/180.*pi
    Blong_rad = Blong/180.*pi

    deltaL = Blong_rad - Along_rad
    X = cos(Blat_rad)*sin(deltaL)
    Y = cos(Alat_rad)*sin(Blat_rad) - sin(Alat_rad)*cos(Blat_rad)*cos(deltaL)

    return arctan2(X, Y)/pi*180. + 180.

def calc_power(v_new, v_old, tdiff, dir, slope, phys_var, verbosity = 0):
    slope_rad = arctan(slope)
    v = 0.5*(v_new + v_old)
    v_wind = v - phys_var['wind_v'] * cos((dir - phys_var['wind_dir'])/180.*pi)

    F_g = phys_var['mass'] * phys_var['g'] * sin(slope_rad)
    F_r = phys_var['mass'] * phys_var['g'] * abs(cos(slope_rad)) * phys_var['crr']
    F_w = 0.5 * phys_var['cda'] * phys_var['rho'] * v_wind*v_wind
    pow_base = (F_g + F_r + F_w) * v

    diff_ekin = 0.5*phys_var['mass']*(v_new**2 - v_old**2)
    diff_erot = 0.5*phys_var['rot_mass']*(v_new**2 - v_old**2)
    pow_acc = (diff_ekin + diff_erot) / tdiff

    if verbosity > 1:
        print('Forces g/r/w: ', F_g, F_r, F_w)
        print('Powers g/r/w/a: ', F_g*v, F_r*v, F_w*v, pow_acc)
        print('Ekin_new: ', 0.5*phys_var['mass']*v_new**2, ' Erot_new: ', 0.5*phys_var['rot_mass']*v_new**2)
        print('Diff Ekin, Diff Erot, Power: ', diff_ekin, diff_erot, pow_base+pow_acc)

    return (pow_base + pow_acc) / (1 - phys_var['loss'])

def calc_power_data(data, phys_var, calc_neg_watts = True, verbosity = 0):
#data needs to contain the following entries:
#'speed', 'power', 'distance', 'position_lat', 'position_long', 'altitude'
#
#Physical parameter dictionary example:
#phys_var = {
#    'mass'     : 73+9,
#    'crr'      : 0.004,
#    'cda'      : 0.215,
#    'rho'      : 1.225,
#    'g'        : 9.81,
#    'loss'     : 0.025,
#    'wind_v'   : 1.,
#    'wind_dir' : -70.
#}
    comp_pow = []
    v_old = data['speed'][0]/3.6
    ndata = len(data['speed'])
    for it in range(ndata):
        v_new = data['speed'][it]/3.6

        # compute time difference
        if it > 0:
            tdiff = data['seconds'][it] - data['seconds'][it-1]
        else:
            tdiff = 1.

        # compute smoothed direction
        if it == ndata - 1: # last step
            dir = calc_direction(data['position_lat'][it-1], data['position_long'][it-1], data['position_lat'][it], data['position_long'][it])
        elif it > 0: # regular step
            dir = calc_direction(data['position_lat'][it-1], data['position_long'][it-1], data['position_lat'][it+1], data['position_long'][it+1])
        else: # first step: assume perfect cross wind
            dir = phys_var['wind_dir'] + 90.

        # compute power
        pow = calc_power(v_new, v_old, tdiff, dir, data['slope'][it], phys_var, verbosity)
        if pow > 0 or calc_neg_watts: # option to clamp power to positive values
            comp_pow.append(pow)
        else:
            comp_pow.append(0.)
        v_old = v_new

    return array(comp_pow)

File: calc/calc_power.c
#include <math.h>
#include <stdio.h>

struct PhysVar {
    const double mass;
    const double rot_mass;
    const double crr;
    const double cda;
    const double rho;
    const double g;
    const double loss;
    const double wind_v;
    const double wind_dir;
};

double calc_direction(const double Alat, const double Along, const double Blat, const double Blong) {
    const double Alat_rad = Alat/180.*M_PI;
    const double Along_rad = Along/180.*M_PI;
    const double Blat_rad = Blat/180.*M_PI;
    const double Blong_rad = Blong/180.*M_PI;

    const double deltaL = Blong_rad - Along_rad;
    const double X = cos(Blat_rad)*sin(deltaL);
    const double Y = cos(Alat_rad)*sin(Blat_rad) - sin(Alat_rad)*cos(Blat_rad)*cos(deltaL);

    return atan2(X, Y)/M_PI*180. + 180.;
}

double calc_power(const double v_new, const double v_old, const double tdiff, const double dir, const double slope, const struct PhysVar * const phys_var) {
    const double slope_rad = atan(slope);
    const double v = 0.5*(v_new + v_old);
    const double v_wind = v - phys_var->wind_v * cos((dir - phys_var->wind_dir)/180.*M_PI);

    const double F_g = phys_var->mass * phys_var->g * sin(slope_rad);
    const double F_r = phys_var->mass * phys_var->g * fabs(cos(slope_rad)) * phys_var->crr;
    const double F_w = 0.5 * phys_var->cda * phys_var->rho * v_wind*v_wind;
    const double pow_base = (F_g + F_r + F_w) * v;

    const double diff_ekin = 0.5*phys_var->mass*(v_new*v_new - v_old*v_old);
    const double diff_erot = 0.5*phys_var->rot_mass*(v_new*v_new - v_old*v_old);
    const double pow_acc = (diff_ekin + diff_erot) / tdiff;

    return (pow_base + pow_acc) / (1. - phys_var->loss);
}

void calc_power_data(const int ndata, double * const comp_pow,
                    const double * const speed, const double * const posLat, const double * const posLong,
                    const double * const slope, const double * const tsecs,
                    const struct PhysVar phys_var, const int calc_neg_watts) {
    double v_old = 0.;
    for (int it = 0; it < ndata; it+=1) {
        const double v_new = speed[it]/3.6;
        double dir;
        double tdiff;
        //printf("Speed: %f, PosLat: %f, PosLong: %f, Slope: %f, Secs: %f", speed[it], posLat[it], posLong[it], slope[it], tsecs[it]);
        //printf("\n");

        // compute time difference
        if (it > 0) {
            tdiff = tsecs[it] - tsecs[it-1];
        }
        else {
            tdiff = 1.;
        }

        // compute smoothed direction
        if (ndata > 1 && it == ndata - 1) { // last step
            dir = calc_direction(posLat[it-1], posLong[it-1], posLat[it], posLong[it]);
        }
        else if (ndata > 1 && it > 0) { // regular step
            dir = calc_direction(posLat[it-1], posLong[it-1], posLat[it+1], posLong[it+1]);
        }
        else { // first step: just assume perfect cross wind
            dir = phys_var.wind_dir + 90.;
        }

        // compute power
        const double pow = calc_power(v_new, v_old, tdiff, dir, slope[it], &phys_var);
        if (pow > 0 || calc_neg_watts > 0) { // option to clamp power to positive values
            comp_pow[it] = pow;
        }
        else {
            comp_pow[it] = 0.;
        }
        v_old = v_new;
    }

    return;
}
